File: packages/native-helpers/linux-helper-ts/src/evdev/keycodes.ts
/** Linux evdev key codes for common keys */

// Modifiers
export const KEY_LEFTCTRL = 29;
export const KEY_LEFTSHIFT = 42;
export const KEY_RIGHTSHIFT = 54;
export const KEY_LEFTALT = 56;
export const KEY_RIGHTALT = 100;
export const KEY_LEFTMETA = 125;
export const KEY_RIGHTMETA = 126;
export const KEY_FN = 464;

// Common keys
export const KEY_V = 47;
export const KEY_C = 46;

const MODIFIER_KEYCODES = new Set([
  KEY_LEFTCTRL,
  KEY_LEFTSHIFT,
  KEY_RIGHTSHIFT,
  KEY_LEFTALT,
  KEY_RIGHTALT,
  KEY_LEFTMETA,
  KEY_RIGHTMETA,
  KEY_FN,
]);

export function isModifierKeyCode(code: number): boolean {
  return MODIFIER_KEYCODES.has(code);
}

export function getModifierFlags(pressedKeys: Set<number>): {
  ctrlKey: boolean;
  shiftKey: boolean;
  altKey: boolean;
  metaKey: boolean;
} {
  return {
    ctrlKey: pressedKeys.has(KEY_LEFTCTRL) || pressedKeys.has(29 + 97), // KEY_RIGHTCTRL=97
    shiftKey: pressedKeys.has(KEY_LEFTSHIFT) || pressedKeys.has(KEY_RIGHTSHIFT),
    altKey: pressedKeys.has(KEY_LEFTALT) || pressedKeys.has(KEY_RIGHTALT),
    metaKey: pressedKeys.has(KEY_LEFTMETA) || pressedKeys.has(KEY_RIGHTMETA),
  };
}

File: packages/native-helpers/linux-helper-py/linux_helper/evdev/keycodes.py
"""Linux evdev key code constants."""

# Modifiers
KEY_LEFTCTRL = 29
KEY_LEFTSHIFT = 42
KEY_RIGHTSHIFT = 54
KEY_LEFTALT = 56
KEY_RIGHTALT = 100
KEY_LEFTMETA = 125
KEY_RIGHTMETA = 126
KEY_RIGHTCTRL = 97
KEY_FN = 464

# Common keys
KEY_V = 47
KEY_C = 46

MODIFIER_KEYCODES = frozenset({
    KEY_LEFTCTRL, KEY_RIGHTCTRL,
    KEY_LEFTSHIFT, KEY_RIGHTSHIFT,
    KEY_LEFTALT, KEY_RIGHTALT,
    KEY_LEFTMETA, KEY_RIGHTMETA,
    KEY_FN,
})


def is_modifier(code: int) -> bool:
    return code in MODIFIER_KEYCODES


def get_modifier_flags(pressed: set[int]) -> dict:
    return {
        "ctrlKey": KEY_LEFTCTRL in pressed or KEY_RIGHTCTRL in pressed,
        "shiftKey": KEY_LEFTSHIFT in pressed or KEY_RIGHTSHIFT in pressed,
        "altKey": KEY_LEFTALT in pressed or KEY_RIGHTALT in pressed,
        "metaKey": KEY_LEFTMETA in pressed or KEY_RIGHTMETA in pressed,
    }
